compiler/src/main/kotlin/ast/ast.kt

```diff
@@ -47,191 +64,252 @@ enum class WorldType {
+class GreaterOrEqExpr(
+    ctx: SourceContext,
+    val left: Expr,
+    val right: Expr
+) : Expr(ctx)
 
-data class LessThanExpr(override val ctx: CellmataParser.LessExprContext, val left: Expr, val right: Expr) : Expr(ctx)
+class LessThanExpr(
+    ctx: SourceContext,
+    val left: Expr,
+    val right: Expr
+) : Expr(ctx)
 
-data class LessEqExpr(override val ctx: CellmataParser.LessEqExprContext, val left: Expr, val right: Expr) : Expr(ctx)
+class LessOrEqExpr(
+    ctx: SourceContext,
+    val left: Expr,
+    val right: Expr
+) : Expr(ctx)
 
-data class AdditionExpr(override val ctx: CellmataParser.AdditionExprContext, val left: Expr, val right: Expr) : Expr(ctx)
+class AdditionExpr(
+    ctx: SourceContext,
+    val left: Expr,
+    val right: Expr
+) : Expr(ctx)
 
-data class SubtractionExpr(override val ctx: CellmataParser.SubstractionExprContext, val left: Expr, val right: Expr) : Expr(ctx)
+class SubtractionExpr(
+    ctx: SourceContext,
+    val left: Expr,
+    val right: Expr
+) : Expr(ctx)
 
-data class MultiplicationExpr(override val ctx: CellmataParser.MultiplictionExprContext, val left: Expr, val right: Expr) : Expr(ctx)
+class MultiplicationExpr(
+    ctx: SourceContext,
+    val left: Expr,
+    val right: Expr
+) : Expr(ctx)
 
-data class DivisionExpr(override val ctx: CellmataParser.DivisionExprContext, val left: Expr, val right: Expr) : Expr(ctx)
+class DivisionExpr(
+    ctx: SourceContext,
+    val left: Expr,
+    val right: Expr
+) : Expr(ctx)
 
-data class NegativeExpr(override val ctx: CellmataParser.NegativeExprContext, val value: Expr) : Expr(ctx)
+class NegationExpr(
+    ctx: SourceContext,
+    val value: Expr
+) : Expr(ctx)
 
-data class InverseExpr(override val ctx: CellmataParser.InverseExprContext, val value: Expr) : Expr(ctx)
+class NotExpr(
+    ctx: SourceContext,
+    val value: Expr
+) : Expr(ctx)
 
-data class ArrayLookupExpr(
-    override val ctx: CellmataParser.ArrayLookupExprContext,
-    val ident: Expr,
+class ArrayLookupExpr(
+    ctx: SourceContext,
+    val arr: Expr,
     val index: Expr
 ) : Expr(ctx)
 
 /**
- * @param ctx Node in the parse tree this node corresponds to.
+ * @param ctx the context from witch this node was created from
  * @param values Elements of the array.
  * @param declaredType Type listed before the body/values of the array.
  */
-data class ArrayBodyExpr(
-    override val ctx: CellmataParser.ArrayValueExprContext,
+class ArrayBodyExpr(
+    ctx: SourceContext,
     val values: List<Expr>,
-    val declaredType: Type? = UncheckedType
+    val declaredType: Type
 ) : Expr(ctx)
 
-data class ParenExpr(override val ctx: CellmataParser.ParenExprContext, val expr: Expr) : Expr(ctx)
-
-data class NamedExpr(override val ctx: ParserRuleContext, var ident: String = MAGIC_UNDEFINED_STRING) : Expr(ctx)
+class Identifier(
+    ctx: SourceContext,
+    var spelling: String
```

Comment:
Why was ident renamed to spelling?

Reply:
Well, NamedExpr are actually just identifiers (names of variables, constants, states, neighbourhoods), so I thought it was weird the names had a name. It also adheres to what we learned in SPO.